### PR TITLE
[Method References 2] LPS-165285

### PR DESCRIPTION
**Pull request by JanisZhang**
Resend. previous P/R:https://github.com/liferay-core-infra/liferay-portal/pull/2381
This PR is to use serviceTrackerMap instead of the method reference in `portal-search-elasticsearch7-impl` . Thanks!

https://issues.liferay.com/browse/LPS-165285

### CODE REVIEW

**Comment by zhangyan0701** [comment]
No regressions found.

**Review by dantewang — APPROVED** [review]

File: modules/apps/portal-search-elasticsearch7/portal-search-elasticsearch7-impl/src/main/java/com/liferay/portal/search/elasticsearch7/internal/facet/CompositeFacetProcessor.java [hunk 1]
```diff
@@ -54,28 +53,15 @@ public Optional<AggregationBuilder> processFacet(Facet facet) {
...
-		String className = MapUtil.getString(properties, "class.name");
-
-		_facetProcessors.remove(className);
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_serviceTrackerMap = ServiceTrackerMapFactory.openSingleValueMap(
+			bundleContext,
+			(Class<FacetProcessor<SearchRequestBuilder>>)
+				(Class<?>)FacetProcessor.class,
+			"(&(class.name=*)(!(class.name=DEFAULT)))",
```
Comment (dantewang) [inline modules/apps/portal-search-elasticsearch7/portal-search-elasticsearch7-impl/src/main/java/com/liferay/portal/search/elasticsearch7/internal/facet/CompositeFacetProcessor.java:62]: @tinatian This class explicitly use OSGi filter to select services with `class.name` property and it's not `DEFAULT`, so I think using the `class.name` property is OK here.

By the way, providing the identifier via method makes it impossible for someone to reference one of the services of an interface. So for

```
public interface A
```

there could be a registry existing:

```
@Component
public class ARegistry
```

but there could also be some components referencing specific instance of A:

```
@Reference(target="(some.property=some.value)")
private A _a;
```

If we meet such case, do you want us to try to eliminate either the registry or the explicit references, or keep the property?

Comment (dantewang) [reply modules/apps/portal-search-elasticsearch7/portal-search-elasticsearch7-impl/src/main/java/com/liferay/portal/search/elasticsearch7/internal/facet/CompositeFacetProcessor.java:62]: @W2jx Do you have any example for the case above?

**Comment by tinatian** [comment]
@JanisZhang As we talked, close, please resend.

Tina

**Pull request closed without merging**